File: MergeSort.ts
function mergeSort(array: number[]): number[] {
    if (array.length <= 1) return array;

    const middleIndex = Math.floor(array.length / 2);
    const left = array.slice(0, middleIndex);
    const right = array.slice(middleIndex);

    const sortedLeft = mergeSort(left);
    const sortedRight = mergeSort(right);

    return merge(sortedLeft, sortedRight);
}

function merge(left: number[], right: number[]): number[] {
    const result: number[] = [];

    let leftIndex = 0;
    let rightIndex = 0;

    while (leftIndex < left.length && rightIndex < right.length) {
        if (left[leftIndex] < right[rightIndex]) {
            result.push(left[leftIndex++]);
        } else {
            result.push(right[rightIndex++]);
        }
    }

    while (leftIndex < left.length) result.push(left[leftIndex++]);
    while (rightIndex < right.length) result.push(right[rightIndex++]);

    return result;
}

function testMergeSort() {
    console.assert(
        JSON.stringify(mergeSort([4, 1, 3, 1, 54, 3, -23])) ===
            JSON.stringify([-23, 1, 1, 3, 3, 4, 54]),
        "Test 1 Failed"
    );
    console.assert(
        JSON.stringify(mergeSort([10, -1, 2, 5, 0, 6, 2, -1, -9, 11])) ===
            JSON.stringify([-9, -1, -1, 0, 2, 2, 5, 6, 10, 11]),
        "Test 2 Failed"
    );
    console.assert(
        JSON.stringify(mergeSort([3, 2, 1])) === JSON.stringify([1, 2, 3]),
        "Test 3 Failed"
    );
    console.assert(
        JSON.stringify(mergeSort([0, 0, 0, 0, 0, -1, -1, -1])) ===
            JSON.stringify([-1, -1, -1, 0, 0, 0, 0, 0]),
        "Test 4 Failed"
    );
    console.assert(
        JSON.stringify(mergeSort([])) === JSON.stringify([]),
        "Test 5 Failed"
    );
    console.assert(
        JSON.stringify(mergeSort([1])) === JSON.stringify([1]),
        "Test 6 Failed"
    );
    console.assert(
        JSON.stringify(mergeSort([2, 1])) === JSON.stringify([1, 2]),
        "Test 7 Failed"
    );
    console.assert(
        JSON.stringify(mergeSort([1, 2])) === JSON.stringify([1, 2]),
        "Test 8 Failed"
    );
}

testMergeSort();

File: Algorithms/Sorting/MergeSort/MergeSort.cs
namespace DSA;

public static class MergeSortImplementation
{
    public static int[] MergeSort(int[] array)
    {
        if (array.Length <= 1) return array;

        var middleIndex = array.Length / 2;
        var left = array[0..middleIndex];
        var right = array[middleIndex..];

        var sortedLeft = MergeSort(left);
        var sortedRight = MergeSort(right);

        return Merge(sortedLeft, sortedRight);
    }

    private static int[] Merge(int[] left, int[] right)
    {
        List<int> result = [];
        var leftIndex = 0;
        var rightIndex = 0;

        while (leftIndex < left.Length && rightIndex < right.Length)
        {
            if (left[leftIndex] < right[rightIndex])
                result.Add(left[leftIndex++]);
            else
                result.Add(right[rightIndex++]);
        }

        while (leftIndex < left.Length)
            result.Add(left[leftIndex++]);

        while (rightIndex < right.Length)
            result.Add(right[rightIndex++]);

        return [.. result];
    }
}
